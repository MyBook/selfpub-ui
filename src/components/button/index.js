import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { ThemeProvider } from "styled-components";
import isNil from "lodash.isnil";

import { Container, Wrapper, Icon, Title, Loader } from "./styles";
import themes from "./../theme/base/button";

class Button extends PureComponent {
  _getCommonButtonProps = () => {
    const {
      className,
      onClick,
      iconName,
      title,
      disabled,
      loading,
      stretched,
    } = this.props;

    const hasIcon = !isNil(iconName);
    const hasTitle = !!title;
    const justIcon = hasIcon && !hasTitle;
    const justTitle = hasTitle && !hasIcon;
    const fitOnAction = loading && !hasIcon;

    return {
      className,
      disabled: disabled || loading,
      onClick: onClick,
      justIcon: justIcon,
      justTitle: justTitle,
      fitOnAction: fitOnAction,
      stretched: stretched,
      loading: loading,
    };
  };

  renderIcon = () => {
    const { iconName, loading } = this.props;
    const hasIcon = !isNil(iconName);

    if (loading) {
      return <Loader />;
    }

    return hasIcon && <Icon name={iconName} />;
  };

  renderChildren = () => {
    const { iconName, iconRight, title } = this.props;

    const hasIcon = !isNil(iconName);
    const hasTitle = !!title;
    const justIcon = hasIcon && !hasTitle;
    const justTitle = hasTitle && !hasIcon;

    return (
      <Wrapper reverse={iconRight} justTitle={justTitle} justIcon={justIcon}>
        {this.renderIcon()}
        {hasTitle && <Title>{title}</Title>}
      </Wrapper>
    );
  };
  renderButton = () => {
    const { type, disabled, href, tabIndex, loading } = this.props;

    if (type === "link") {
      const linkDisabled = disabled || loading;

      return (
        <Container
          {...this._getCommonButtonProps()}
          as="a"
          role="button"
          href={linkDisabled ? undefined : href}
          tabIndex={linkDisabled ? -1 : tabIndex}
          aria-disabled={linkDisabled}
        >
          {this.renderChildren()}
        </Container>
      );
    }

    return (
      <Container {...this._getCommonButtonProps()}>
        {this.renderChildren()}
      </Container>
    );
  };

  render() {
    const { variation, appearance } = this.props;

    const appearanceTheme = themes[appearance] || themes.common;
    const { base = themes.common.base } = appearanceTheme;
    const variationTheme =
      appearanceTheme[variation || "default"] || themes.common.default;
    const theme = { ...variationTheme, base };

    return <ThemeProvider theme={theme}>{this.renderButton()}</ThemeProvider>;
  }
}

function requiredPropsCheck(props, propName, componentName) {
  if (isNil(props.iconName) && isNil(props.title)) {
    return new Error(
      `One of props 'iconName' or 'title' was not specified in '${componentName}'.`,
    );
  }

  if (!isNil(props[propName]) && typeof props[propName] !== "string") {
    return new Error(
      `Invalid prop '${propName}' supplied to '${componentName}', expected 'string'.`,
    );
  }

  if (!isNil(props[propName]) && props[propName].length === 0) {
    return new Error(
      `Invalid prop '${propName}' supplied to '${componentName}', expected not empty string.`,
    );
  }
}

function requiredHref({ type, href }, propName, componentName) {
  if (type === "link" && !href) {
    return new Error(
      `Then 'href' prop was not specified for type link in '${componentName}'.`,
    );
  }

  if (type !== "link" && href) {
    if (typeof href !== "string") {
      return new Error(
        `Invalid prop '${propName}' supplied to '${componentName}', expected 'string'.`,
      );
    }

    return new Error(
      `The 'href' prop should be used only for type link in '${componentName}'.`,
    );
  }
}

function requiredOnClick({ type, onClick }, propName, componentName) {
  if (type !== "link" && !onClick) {
    return new Error(
      `Then 'onClick' prop was not specified for type button in '${componentName}'.`,
    );
  }

  if (typeof onClick !== "function") {
    return new Error(
      `Invalid prop '${propName}' supplied to '${componentName}', expected 'func'.`,
    );
  }
}

Button.propTypes = {
  className: PropTypes.string,
  type: PropTypes.string,
  onClick: requiredOnClick,
  iconName: requiredPropsCheck,
  title: requiredPropsCheck,
  iconRight: PropTypes.bool,
  variation: PropTypes.oneOf(Object.keys(themes.flat)),
  appearance: PropTypes.oneOf(Object.keys(themes)),
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  stretched: PropTypes.bool,
  href: requiredHref,
  tabIndex: PropTypes.number,
};

Button.defaultProps = {
  className: "",
  type: "button",
  iconRight: false,
  variation: "default",
  appearance: "raised",
  disabled: false,
  loading: false,
  stretched: true,
  tabIndex: 0,
};

export default Button;
