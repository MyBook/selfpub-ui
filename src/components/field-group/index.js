export { default } from "./field-group";
