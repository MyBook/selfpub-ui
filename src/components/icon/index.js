import React from "react";
import PropTypes from "prop-types";

const Icon = ({ name, size, color, className }) => (
  <svg className={className} fill={color} height={size} width={size}>
    <use xlinkHref={`#${name}`} />
  </svg>
);

Icon.displayName = "Icon";

Icon.propTypes = {
  name: PropTypes.string.isRequired,
  size: PropTypes.number,
  color: PropTypes.string,
  className: PropTypes.string,
};

Icon.defaultProps = {
  className: "",
  color: "#000",
  size: 24,
};

export default Icon;
