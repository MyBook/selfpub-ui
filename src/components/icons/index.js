export * as Icon from "icons/index";
