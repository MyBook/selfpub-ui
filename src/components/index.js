/* global process */
export { default as Button } from "./button/index";
export { default as Checkbox } from "./checkbox/index";
export { default as Radio } from "./radio/index";
export { default as RadioGroup } from "./radio-group/index";
export { default as InfoBox } from "./info-box/index";
export { default as FieldGroup } from "./field-group/index";
export { default as Icon } from "./icons/index";
export { default as InlineSelect } from "./inline-select/index";
export { default as Select } from "./select/index";
export { default as Link } from "./link/index";
export { default as Loader } from "./loader/loader";
export { default as Popover } from "./popover/index";
export { default as Textarea } from "./textarea/index";
export { ThemeProviderWrapper, colors } from "./theme/index";
export { Typography, Header, Paragraph } from "./typography/index";
export { Toggle, ToggleOption } from "./toggle/index";
export {
  Input,
  InputStatus,
  ErrorField,
  LabelWrapper,
  InputWrapper,
} from "./input/index";
export { Table, Td, Tr } from "./table/index";
export { Tab, Tabs } from "./tabs/index";
export { default as SearchBox } from "./search-box/index";
