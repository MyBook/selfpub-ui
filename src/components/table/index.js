export { Table, Td, Tr } from "./table";
