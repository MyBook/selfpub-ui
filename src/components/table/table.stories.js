/* global module */
import React from "react";
import { storiesOf } from "@storybook/react";
import withTestsWrapper from "../../../.storybook/withTest";
import { TableExample } from "./__stories__/all-in-one";

storiesOf("Table", module)
  .addDecorator(withTestsWrapper("table"))
  .add("1. All in one", () => <TableExample />);
