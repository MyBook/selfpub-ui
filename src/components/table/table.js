import React, { PureComponent } from "react";
import styled from "styled-components";
import { mainTheme } from "../../components/ui-styles";
import { lighten } from "../../utils/styled";

const StyledBaseTabelItem = styled.div`
  padding: 10px 20px;
  color: ${mainTheme.color.coal};
  font: 20px/1em;
  text-align: left;
`;

export class Table extends PureComponent {
  render() {
    const { children } = this.props;

    const StyledTable = styled.div`
      color: ${mainTheme.color.coal};
    `;

    return <StyledTable>{children}</StyledTable>;
  }
}

export class Tr extends PureComponent {
  render() {
    const { children, isHeader, onClick } = this.props;

    const StyledTr = styled.div`
      display: flex;
      align-items: center;
      font-size: 17px;
      font-weight: 500;
      transition: 0.2s;
      opacity: 1;
      min-height: 80px;
      white-space: nowrap;
      padding: 0 10px;

      &:nth-child(odd) {
        background: ${lighten(mainTheme.color.strongLighterCoal, 20)};
      }

      :first-child {
        background: ${mainTheme.color.beige};
        font-weight: 400;
      }
    `;

    return <StyledTr onClick={onClick}>{children}</StyledTr>;
  }
}

export class Td extends PureComponent {
  render() {
    const { children, onClick, style } = this.props;

    const StyledTd = styled.div`
      display: flex;
      align-items: center;
      width: 18%;
      white-space: nowrap;
      padding: 0 10px;
    `;

    return (
        <StyledTd onClick={onClick} style={style}>
        {children}
      </StyledTd>
    );
  }
}
