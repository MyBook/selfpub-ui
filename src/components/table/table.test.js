import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";
import { Table, Tr, Td } from "../../components/table";

test("Table is rendered", () => {
  const table = render(
    <Table>
      <Tr>
        <Td style={{ width: "200px" }}>123</Td>
      </Tr>
    </Table>,
  );

  expect(toJson(table)).toMatchSnapshot();
});
