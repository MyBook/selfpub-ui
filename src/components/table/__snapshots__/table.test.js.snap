// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table is rendered 1`] = `
<div
  class="sc-bwzfXH iNkJQj"
>
  <div
    class="sc-htpNat bZGgon"
  >
    <div
      class="sc-bxivhb cZnznB"
      style="width:200px"
    >
      123
    </div>
  </div>
</div>
`;
